### PR TITLE
exposing provider metrics in controller

### PR DESCRIPTION
**Pull request by laurafitzgerald**
Co-authored-by: ciaranRoche <croche@redhat.com>

## Overview

Jira: https://issues.redhat.com/browse/INTLY-7226
Jira: https://issues.redhat.com/browse/INTLY-7229
Jira: https://issues.redhat.com/browse/INTLY-7230

## Verification

- Clone this branch
- Run `make cluster/prepare`
- Run `make cluster/seed/managed/postgres` - do this twice if you like. :D 
- Run `make run`
- Run `curl localhost:8383/metrics | grep cro_postgres_`
- You should see output similar to: 
```
╭─croche@localhost ~/Projects/integreatly/cloud-resource-operator ‹intly-7226› 
╰─$ curl localhost:8383/metrics | grep cro_postgres_                                                                                                                                                                             130 ↵

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
# HELP cro_postgres_cpu_utilization_average The percentage of CPU utilization. Units: Percent
# TYPE cro_postgres_cpu_utilization_average gauge
cro_postgres_cpu_utilization_average{clusterID="hcd-rhmi-clouds-6pdvc",instanceID="hcdrhmiclouds6pdvccloudresourceoper-eu6t",namespace="cloud-resource-operator",productName="productName",resourceID="example-postgres",strategy="aws-rds"} 1.5
cro_postgres_cpu_utilization_average{clusterID="hcd-rhmi-clouds-6pdvc",instanceID="hcdrhmiclouds6pdvccloudresourceoper-zaj2",namespace="cloud-resource-operator",productName="productName",resourceID="example-postgres-metric",strategy="aws-rds"} 1.8
# HELP cro_postgres_free_storage_average The amount of available storage space. Units: Bytes
# TYPE cro_postgres_free_storage_average gauge
cro_postgres_free_storage_average{clusterID="hcd-rhmi-clouds-6pdvc",instanceID="hcdrhmiclouds6pdvccloudresourceoper-eu6t",namespace="cloud-resource-operator",productName="productName",resourceID="example-postgres",strategy="aws-rds"} 2.00888623104e+10
cro_postgres_free_storage_average{clusterID="hcd-rhmi-clouds-6pdvc",instanceID="hcdrhmiclouds6pdvccloudresourceoper-zaj2",namespace="cloud-resource-operator",productName="productName",resourceID="example-postgres-metric",strategy="aws-rds"} 2.00888614912e+10
# HELP cro_postgres_freeable_memory_average The amount of available random access memory. Units: Bytes
# TYPE cro_postgres_freeable_memory_average gauge
cro_postgres_freeable_memory_average{clusterID="hcd-rhmi-clouds-6pdvc",instanceID="hcdrhmiclouds6pdvccloudresourceoper-eu6t",namespace="cloud-resource-operator",productName="productName",resourceID="example-postgres",strategy="aws-rds"} 9.495584768e+08
cro_postgres_freeable_memory_average{clusterID="hcd-rhmi-clouds-6pdvc",instanceID="hcdrhmiclouds6pdvccloudresourceoper-zaj2",namespace="cloud-resource-operator",productName="productName",resourceID="example-postgres-metric",strategy="aws-rds"} 9.441558528e+08
100 33247    0 33247    0     0  6493k      0 --:--:-- --:--:-- --:--:-- 6493k


```

### CODE REVIEW

File: pkg/controller/cloudmetrics/cloudmetrics_controller.go [hunk 1]
```diff
@@ -115,7 +216,7 @@ func (r *ReconcileCloudMetrics) Reconcile(request reconcile.Request) (reconcile.
 			}
 
 			for _, metricData := range scrapedMetricsOutput.Metrics {
-				r.logger.Debug(*metricData)
+				scrapedMetrics = append(scrapedMetrics, metricData)
 			}
```
Comment (aidenkeating) [inline pkg/controller/cloudmetrics/cloudmetrics_controller.go:220]: `scrapedMetrics = append(scrapedMetrics, scrapedMetricsOutput.Metrics...)`

File: pkg/controller/cloudmetrics/cloudmetrics_controller.go [hunk 2]
```diff
@@ -127,23 +228,52 @@ func (r *ReconcileCloudMetrics) Reconcile(request reconcile.Request) (reconcile.
...
 			if err != nil {
 				r.logger.Errorf("failed to scrape metrics for postgres %v", err)
 				continue
 			}
 
+			// add the returned scraped metrics to the list of metrics
 			for _, metricData := range scrapedMetricsOutput.Metrics {
-				r.logger.Debug(*metricData)
+				scrapedMetrics = append(scrapedMetrics, metricData)
+			}
```
Comment (aidenkeating) [inline pkg/controller/cloudmetrics/cloudmetrics_controller.go:260]: `scrapedMetrics = append(scrapedMetrics, scrapedMetricsOutput.Metrics...)`

File: pkg/controller/cloudmetrics/cloudmetrics_controller.go [hunk 3]
```diff
@@ -17,10 +19,98 @@ import (
...
+				Help: "The amount of available storage space. Units: Bytes",
+			},
+			labels),
+		ProviderType: map[string]providers.CloudProviderMetricType{
+			providers.AWSDeploymentStrategy: {
+				PromethuesMetricName: postgresFreeStorageAverage,
+				ProviderMetricName:   "FreeStorageSpace",
+				Statistic:            cloudwatch.StatisticAverage,
+			},
+		},
```
Comment (aidenkeating) [inline pkg/controller/cloudmetrics/cloudmetrics_controller.go:74]: I'm trying to understand this approach, do you have time to discuss this sometime today?

**Comment by psturc** [comment]
👍  verified locally

File: pkg/controller/cloudmetrics/cloudmetrics_controller.go (hunk 3 above)
Comment (ciaranRoche) [reply pkg/controller/cloudmetrics/cloudmetrics_controller.go:74]: sure, let us know when you are free :+1:

**Pull request merged**